File: packages/python/elizaos/types/runtime.py
from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import AsyncIterator, Awaitable, Callable
from typing import TYPE_CHECKING, Any

from pydantic import BaseModel, Field

from elizaos.logger import Logger
from elizaos.types.database import IDatabaseAdapter
from elizaos.types.primitives import UUID, Content

if TYPE_CHECKING:
    from elizaos.types.agent import Character, TemplateType
    from elizaos.types.components import (
        Action,
        ActionResult,
        Evaluator,
        HandlerCallback,
        Provider,
    )
    from elizaos.types.environment import Entity, Room, World
    from elizaos.types.memory import Memory
    from elizaos.types.model import (
        GenerateTextOptions,
        GenerateTextResult,
        ModelType,
    )
    from elizaos.types.plugin import Plugin, Route
    from elizaos.types.service import Service
    from elizaos.types.state import State
    from elizaos.types.task import TaskWorker


# Type alias for streaming model handlers
StreamingModelHandler = Callable[["IAgentRuntime", dict[str, Any]], AsyncIterator[str]]


# Runtime settings type
RuntimeSettings = dict[str, str | bool | int | float | None]


# Send handler function type
SendHandlerFunction = Callable[[Any, Content], Awaitable[None]]


class TargetInfo(BaseModel):
    room_id: UUID | None = Field(default=None, alias="roomId")
    entity_id: UUID | None = Field(default=None, alias="entityId")
    world_id: UUID | None = Field(default=None, alias="worldId")
    source: str | None = None

    model_config = {"populate_by_name": True, "extra": "allow"}


class IAgentRuntime(ABC):
    # Properties that must be implemented
    @property
    @abstractmethod
    def agent_id(self) -> UUID: ...

    @property
    @abstractmethod
    def character(self) -> Character: ...

    @property
    @abstractmethod
    def providers(self) -> list[Provider]: ...

    @property
    @abstractmethod
    def actions(self) -> list[Action]: ...

    @property
    @abstractmethod
    def evaluators(self) -> list[Evaluator]: ...

    @property
    @abstractmethod
    def plugins(self) -> list[Plugin]: ...

    @property
    @abstractmethod
    def services(self) -> dict[str, list[Service]]: ...

    @property
    @abstractmethod
    def routes(self) -> list[Route]: ...

    @property
    @abstractmethod
    def events(self) -> dict[str, list[Callable[[Any], Awaitable[None]]]]: ...

    @property
    @abstractmethod
    def state_cache(self) -> dict[str, State]: ...

    @property
    @abstractmethod
    def message_service(self) -> Any | None: ...

    @property
    @abstractmethod
    def enable_autonomy(self) -> bool: ...

    @enable_autonomy.setter
    @abstractmethod
    def enable_autonomy(self, value: bool) -> None: ...

    # Database adapter
    @abstractmethod
    def register_database_adapter(self, adapter: IDatabaseAdapter) -> None: ...

    @abstractmethod
    async def get_connection(self) -> Any: ...

    @abstractmethod
    async def get_cache(self, key: str) -> object | None: ...

    @abstractmethod
    async def set_cache(self, key: str, value: object) -> bool: ...

    @abstractmethod
    async def delete_cache(self, key: str) -> None: ...

    @abstractmethod
    async def get_memories(
        self, params: dict[str, Any] | None = None, **kwargs: Any
    ) -> list[Any]: ...

    @abstractmethod
    async def create_memory(
        self,
        memory: dict[str, object] | None = None,
        table_name: str | None = None,
        unique: bool | None = None,
        **kwargs: object,
    ) -> Any: ...

    # Plugin management
    @abstractmethod
    async def register_plugin(self, plugin: Plugin) -> None: ...

    @abstractmethod
    async def initialize(
        self, config: dict[str, str | int | bool | None] | None = None
    ) -> None: ...

    # Service management
    @abstractmethod
    def get_service(self, service: str) -> Service | None: ...

    @abstractmethod
    def get_services_by_type(self, service: str) -> list[Service]: ...

    @abstractmethod
    def get_all_services(self) -> dict[str, list[Service]]: ...

    @abstractmethod
    async def register_service(self, service: type[Service]) -> None: ...

    @abstractmethod
    async def get_service_load_promise(self, service_type: str) -> Service: ...

    @abstractmethod
    def get_registered_service_types(self) -> list[str]: ...

    @abstractmethod
    def has_service(self, service_type: str) -> bool: ...

    @abstractmethod
    async def enable_knowledge(self) -> None: ...

    @abstractmethod
    async def disable_knowledge(self) -> None: ...

    @abstractmethod
    def is_knowledge_enabled(self) -> bool: ...

    @abstractmethod
    async def enable_relationships(self) -> None: ...

    @abstractmethod
    async def disable_relationships(self) -> None: ...

    @abstractmethod
    def is_relationships_enabled(self) -> bool: ...

    @abstractmethod
    async def enable_trajectories(self) -> None: ...

    @abstractmethod
    async def disable_trajectories(self) -> None: ...

    @abstractmethod
    def is_trajectories_enabled(self) -> bool: ...

    # Settings
    @abstractmethod
    def set_setting(self, key: str, value: object | None, secret: bool = False) -> None: ...

    @abstractmethod
    def get_setting(self, key: str) -> object | None: ...

    @abstractmethod
    def get_all_settings(self) -> dict[str, object | None]: ...

    @abstractmethod
    def compose_prompt(self, *, state: State, template: TemplateType) -> str: ...

    @abstractmethod
    def compose_prompt_from_state(self, *, state: State, template: TemplateType) -> str: ...

    @abstractmethod
    def get_current_time_ms(self) -> int: ...

    @abstractmethod
    def get_conversation_length(self) -> int: ...

    @property
    @abstractmethod
    def logger(self) -> Logger: ...

    @abstractmethod
    def is_action_planning_enabled(self) -> bool: ...

    @abstractmethod
    def is_check_should_respond_enabled(self) -> bool: ...

    # Action processing
    @abstractmethod
    async def process_actions(
        self,
        message: Memory,
        responses: list[Memory],
        state: State | None = None,
        callback: HandlerCallback | None = None,
        options: dict[str, Any] | None = None,
    ) -> None: ...

    @abstractmethod
    def get_action_results(self, message_id: UUID) -> list[ActionResult]: ...

    # Evaluation
    @abstractmethod
    async def evaluate(
        self,
        message: Memory,
        state: State | None = None,
        did_respond: bool = False,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> list[Evaluator] | None: ...

    # Component registration
    @abstractmethod
    def register_provider(self, provider: Provider) -> None: ...

    @abstractmethod
    def register_action(self, action: Action) -> None: ...

    @abstractmethod
    def register_evaluator(self, evaluator: Evaluator) -> None: ...

    # Connection management
    @abstractmethod
    async def ensure_connections(
        self,
        entities: list[Entity],
        rooms: list[Room],
        source: str,
        world: World,
    ) -> None: ...

    @abstractmethod
    async def ensure_connection(
        self,
        entity_id: UUID,
        room_id: UUID,
        world_id: UUID,
        user_name: str | None = None,
        name: str | None = None,
        world_name: str | None = None,
        source: str | None = None,
        channel_id: str | None = None,
        message_server_id: UUID | None = None,
        channel_type: str | None = None,
        user_id: UUID | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> None: ...

    @abstractmethod
    async def ensure_participant_in_room(self, entity_id: UUID, room_id: UUID) -> None: ...

    @abstractmethod
    async def ensure_world_exists(self, world: World) -> None: ...

    @abstractmethod
    async def ensure_room_exists(self, room: Room) -> None: ...

    # State composition
    @abstractmethod
    async def compose_state(
        self,
        message: Memory,
        include_list: list[str] | None = None,
        only_include: bool = False,
        skip_cache: bool = False,
    ) -> State: ...

    # Model usage
    @abstractmethod
    def has_model(self, model_type: str | ModelType) -> bool: ...

    @abstractmethod
    async def use_model(
        self,
        model_type: str | ModelType,
        params: dict[str, Any] | None = None,
        provider: str | None = None,
        **kwargs: Any,
    ) -> Any: ...

    @abstractmethod
    async def generate_text(
        self,
        input_text: str,
        options: GenerateTextOptions | None = None,
    ) -> GenerateTextResult: ...

    @abstractmethod
    def register_model(
        self,
        model_type: str | ModelType,
        handler: Callable[[IAgentRuntime, dict[str, Any]], Awaitable[Any]],
        provider: str,
        priority: int = 0,
    ) -> None: ...

    @abstractmethod
    def get_model(
        self, model_type: str
    ) -> Callable[[IAgentRuntime, dict[str, Any]], Awaitable[Any]] | None: ...

    @abstractmethod
    def use_model_stream(
        self,
        model_type: str | ModelType,
        params: dict[str, Any] | None = None,
        provider: str | None = None,
        **kwargs: Any,
    ) -> AsyncIterator[str]:
        """
        Use a streaming model handler to generate text token by token.

        Args:
            model_type: The model type (e.g., ModelType.TEXT_LARGE_STREAM)
            params: Parameters for the model (prompt, system, temperature, etc.)
            provider: Optional specific provider to use
            **kwargs: Additional parameters merged into params

        Returns:
            An async iterator yielding text chunks as they are generated.
        """
        ...

    @abstractmethod
    def register_streaming_model(
        self,
        model_type: str | ModelType,
        handler: StreamingModelHandler,
        provider: str,
        priority: int = 0,
    ) -> None:
        """Register a streaming model handler."""
        ...

    # Event handling
    @abstractmethod
    def register_event(
        self,
        event: str,
        handler: Callable[[Any], Awaitable[None]],
    ) -> None: ...

    @abstractmethod
    def get_event(self, event: str) -> list[Callable[[Any], Awaitable[None]]] | None: ...

    @abstractmethod
    async def emit_event(
        self,
        event: str | list[str],
        params: Any,
    ) -> None: ...

    # Task management
    @abstractmethod
    def register_task_worker(self, task_handler: TaskWorker) -> None: ...

    @abstractmethod
    def get_task_worker(self, name: str) -> TaskWorker | None: ...

    # Lifecycle
    @abstractmethod
    async def stop(self) -> None: ...

    # Memory/embedding helpers
    @abstractmethod
    async def add_embedding_to_memory(self, memory: Memory) -> Memory: ...

    @abstractmethod
    async def queue_embedding_generation(
        self, memory: Memory, priority: str = "normal"
    ) -> None: ...

    @abstractmethod
    async def get_all_memories(self) -> list[Memory]: ...

    @abstractmethod
    async def clear_all_agent_memories(self) -> None: ...

    @abstractmethod
    async def update_memory(self, memory: Memory | dict[str, Any]) -> bool: ...

    @abstractmethod
    async def delete_memory(self, memory_id: UUID) -> None: ...

    # Run tracking
    @abstractmethod
    def create_run_id(self) -> UUID: ...

    @abstractmethod
    def start_run(self, room_id: UUID | None = None) -> UUID: ...

    @abstractmethod
    def end_run(self) -> None: ...

    @abstractmethod
    def get_current_run_id(self) -> UUID: ...

    # Convenience wrappers
    @abstractmethod
    async def get_entity_by_id(self, entity_id: UUID) -> Entity | None: ...

    @abstractmethod
    async def get_entity(self, entity_id: UUID | str) -> Entity | None: ...

    @abstractmethod
    async def update_entity(self, entity: Entity) -> None: ...

    @abstractmethod
    async def get_room(self, room_id: UUID) -> Room | None: ...

    @abstractmethod
    async def create_entity(self, entity: Entity) -> bool: ...

    @abstractmethod
    async def get_component(
        self,
        entity_id: UUID,
        component_type: str,
        world_id: UUID | str | None = None,
        source_entity_id: UUID | None = None,
    ) -> Any | None: ...

    @abstractmethod
    async def get_components(
        self,
        entity_id: UUID,
        world_id: UUID | str | None = None,
        source_entity_id: UUID | None = None,
    ) -> list[Any]: ...

    @abstractmethod
    async def create_component(self, component: Any) -> bool: ...

    @abstractmethod
    async def set_component(
        self,
        entity_id: UUID,
        component_type: str,
        data: dict[str, Any],
        room_id: UUID | None = None,
        world_id: UUID | None = None,
        source_entity_id: UUID | None = None,
    ) -> bool: ...

    @abstractmethod
    async def update_component(self, component: Any) -> None: ...

    @abstractmethod
    async def delete_component(
        self, component_id: UUID, component_type: str | None = None
    ) -> None: ...

    @abstractmethod
    async def create_room(self, room: Room) -> UUID: ...

    @abstractmethod
    async def add_participant(self, entity_id: UUID, room_id: UUID) -> bool: ...

    @abstractmethod
    async def get_rooms(self, world_id: UUID) -> list[Room]: ...

    @abstractmethod
    def register_send_handler(self, source: str, handler: SendHandlerFunction) -> None: ...

    @abstractmethod
    async def send_message_to_target(self, target: TargetInfo, content: Content) -> None: ...

    @abstractmethod
    async def update_world(self, world: World) -> None: ...

    @abstractmethod
    async def get_world(self, world_id: UUID) -> World | None: ...

    @abstractmethod
    async def get_relationships(self, params: dict[str, object]) -> list[object]: ...

    @abstractmethod
    async def get_relationships_by_pairs(
        self, pairs: list[dict[str, str]]
    ) -> list[object | None]: ...

    @abstractmethod
    async def create_relationships(self, relationships: list[dict[str, Any]]) -> list[str]: ...

    @abstractmethod
    async def get_relationships_by_ids(self, relationship_ids: list[str]) -> list[object]: ...

    @abstractmethod
    async def update_relationships(self, relationships: list[object]) -> None: ...

    @abstractmethod
    async def delete_relationships(self, relationship_ids: list[str]) -> None: ...

    @abstractmethod
    async def search_knowledge(self, query: str, limit: int = 5) -> list[object]: ...

    @abstractmethod
    async def search_memories(self, params: dict[str, Any]) -> list[Any]: ...

    @abstractmethod
    async def get_entities_for_room(
        self, room_id: UUID, include_components: bool = False
    ) -> list[Any]: ...

    @abstractmethod
    async def update_room(self, room: Any) -> None: ...

File: packages/rust/src/features/basic_capabilities/bootstrap/runtime.rs
//! Runtime trait definitions for the elizaOS plugin system.
//!
//! This module defines the traits that the agent runtime must implement
//! for plugins to interact with the core system.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;
use uuid::Uuid;

use crate::error::PluginResult;
use crate::types::task::Task;
use crate::types::{Character, Content, Entity, Memory, MemoryType, ModelType, Room, State, World};

/// The agent runtime interface.
///
/// This trait defines the contract between plugins and the core runtime.
/// Plugins use this interface to access agent capabilities.
#[async_trait]
pub trait IAgentRuntime: Send + Sync {
    /// Get the agent's unique identifier.
    fn agent_id(&self) -> Uuid;

    /// Get the agent's character definition.
    fn character(&self) -> Character;

    /// Get a setting value.
    fn get_setting(&self, key: &str) -> Option<String>;

    /// Get all settings.
    fn get_all_settings(&self) -> HashMap<String, String>;

    /// Set a setting value.
    async fn set_setting(&self, key: &str, value: &str) -> PluginResult<()>;

    /// Get an entity by ID.
    async fn get_entity(&self, entity_id: Uuid) -> PluginResult<Option<Entity>>;

    /// Update an entity.
    async fn update_entity(&self, entity: &Entity) -> PluginResult<()>;

    /// Get a room by ID.
    async fn get_room(&self, room_id: Uuid) -> PluginResult<Option<Room>>;

    /// Get a world by ID.
    async fn get_world(&self, world_id: Uuid) -> PluginResult<Option<World>>;

    /// Update a world.
    async fn update_world(&self, world: &World) -> PluginResult<()>;

    /// Create a memory entry.
    async fn create_memory(
        &self,
        content: Content,
        room_id: Option<Uuid>,
        entity_id: Option<Uuid>,
        memory_type: MemoryType,
        metadata: HashMap<String, serde_json::Value>,
    ) -> PluginResult<Memory>;

    /// Get memories with filters.
    async fn get_memories(
        &self,
        room_id: Option<Uuid>,
        entity_id: Option<Uuid>,
        memory_type: Option<MemoryType>,
        limit: usize,
    ) -> PluginResult<Vec<Memory>>;

    /// Search the knowledge base.
    async fn search_knowledge(&self, query: &str, limit: usize) -> PluginResult<Vec<Memory>>;

    /// Compose state for a message.
    async fn compose_state(&self, message: &Memory, providers: &[&str]) -> PluginResult<State>;

    /// Compose a prompt from state and template.
    fn compose_prompt(&self, state: &State, template: &str) -> String;

    /// Use a model for inference.
    async fn use_model(
        &self,
        model_type: ModelType,
        params: ModelParams,
    ) -> PluginResult<ModelOutput>;

    /// Check if a model type is available.
    fn has_model(&self, model_type: ModelType) -> bool;

    /// Get available actions.
    fn get_available_actions(&self) -> Vec<ActionInfo>;

    /// Get the current timestamp.
    fn get_current_timestamp(&self) -> i64;

    /// Log an info message.
    fn log_info(&self, source: &str, message: &str);

    /// Log a debug message.
    fn log_debug(&self, source: &str, message: &str);

    /// Log a warning message.
    fn log_warning(&self, source: &str, message: &str);

    /// Log an error message.
    fn log_error(&self, source: &str, message: &str);

    // =========================================================================
    // Task Worker Methods (parity with TypeScript)
    // =========================================================================

    /// Register a task worker.
    fn register_task_worker(&self, worker: Box<dyn TaskWorker>);

    /// Get a task worker by name.
    fn get_task_worker(&self, name: &str) -> Option<Arc<dyn TaskWorker>>;

    // =========================================================================
    // Task CRUD Methods
    // =========================================================================

    /// Create a new task.
    async fn create_task(&self, task: Task) -> PluginResult<Task>;

    /// Get tasks matching the given filters.
    async fn get_tasks(&self, tags: Option<Vec<String>>) -> PluginResult<Vec<Task>>;

    /// Delete a task by ID.
    async fn delete_task(&self, task_id: Uuid) -> PluginResult<bool>;

    /// Get the service for the given type.
    fn get_service(&self, service_type: &str) -> Option<Arc<dyn std::any::Any + Send + Sync>>;
}

/// Task worker trait - defines the contract for executing tasks.
/// Parity with TypeScript's TaskWorker interface.
#[async_trait]
pub trait TaskWorker: Send + Sync {
    /// The unique name of the task type this worker handles
    fn name(&self) -> &str;

    /// Execute the task
    async fn execute(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        options: HashMap<String, serde_json::Value>,
        task: Task,
    ) -> PluginResult<()>;

    /// Optional validation function (defaults to true)
    async fn validate(
        &self,
        _runtime: Arc<dyn IAgentRuntime>,
        _message: Memory,
        _state: State,
    ) -> bool {
        true
    }
}

/// Parameters for model calls.
#[derive(Debug, Clone)]
pub struct ModelParams {
    /// Text prompt for the model
    pub prompt: Option<String>,
    /// Text to embed (for embedding models)
    pub text: Option<String>,
    /// Additional parameters
    pub options: HashMap<String, serde_json::Value>,
}

impl ModelParams {
    /// Create params with just a prompt.
    pub fn with_prompt(prompt: impl Into<String>) -> Self {
        Self {
            prompt: Some(prompt.into()),
            text: None,
            options: HashMap::new(),
        }
    }

    /// Create params with just text (for embeddings).
    pub fn with_text(text: impl Into<String>) -> Self {
        Self {
            prompt: None,
            text: Some(text.into()),
            options: HashMap::new(),
        }
    }
}

/// Output from a model call.
#[derive(Debug, Clone)]
pub enum ModelOutput {
    /// Text output
    Text(String),
    /// Embedding vector
    Embedding(Vec<f32>),
    /// Image URL
    ImageUrl(String),
    /// Structured output
    Structured(serde_json::Value),
}

impl ModelOutput {
    /// Get as text, if this is a text output.
    pub fn as_text(&self) -> Option<&str> {
        match self {
            Self::Text(s) => Some(s),
            _ => None,
        }
    }

    /// Get as embedding, if this is an embedding output.
    pub fn as_embedding(&self) -> Option<&[f32]> {
        match self {
            Self::Embedding(v) => Some(v),
            _ => None,
        }
    }
}

/// Information about an available action.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ActionInfo {
    /// Action name
    pub name: String,
    /// Action description
    pub description: String,
}
